lib/bats-core/test_functions.bash

```diff
@@ -179,6 +179,176 @@ bats_separate_lines() { # <output-array> <input-var>
+      local associated_pipe_position="${pipe_positions[$associated_pipe]}"
+      command_position="$(( $associated_pipe_position + 1 ))"
+    fi
+    local command_to_run="${commands_and_args[$command_position]}"
+
+    local arguments_position="$(( $command_position + 1 ))"
+    local next_pipe_position="${pipe_positions[$given_position]}"
+    local arg_count="$(( next_pipe_position - command_position - 1 ))"
+
+    if [ "$arg_count" -eq 0 ]; then
+      $command_to_run
+    else
+      local arguments=("${commands_and_args[@]:$arguments_position:$arg_count}")
+      $command_to_run "${arguments[@]}"
+    fi
+  }
+
+  # recursive func to handle piping
+  # will run all commands before the given position, chaining piping.
+  # Ex: if "3" is given, this will run `command0 | command1 | command2`.
+  run_leading_commands_until_position() {
+    local exclusive_position="$1"
+
+    local result_status=
+    if [ "$exclusive_position" -le 1 ]; then
+      run_command_at_position 0
+      result_status="$?"
+    else
+      local last_position_to_run="$(( exclusive_position - 1 ))"
+
+      # recurse by calling earlier commands which will then be piped into the
+      # (local) last command to run.
+      run_leading_commands_until_position "$last_position_to_run" | run_command_at_position "$last_position_to_run"
+      # note that we are immediately grabbing PIPESTATUS, and not grabbing $?
+      # we can only grab one of the two, but we can get the equivalent of $?
+      # from PIPESTATUS (see below).
+      local pipe_status=("${PIPESTATUS[@]}")
+
+      # check if we are to return the status code from an exact position, or
+      # use "last fail" (like `set -o pipefail` would).
+      if [ ! -z "$pipefail_position" ]; then
+        # if the target pipefail is less than our "last_position_to_run", then
+        # it is being propagated through the left-hand command.
+        if [ "$pipefail_position" -lt "$last_position_to_run" ]; then
+          result_status="${pipe_status[0]}"
+        elif [ "$pipefail_position" -eq "$last_position_to_run" ]; then
+          # if the target pipefail is equal to our "last_position_to_run", then
+          # it is being returned through the right-hand command.
+          result_status="${pipe_status[1]}"
+        else
+          # if the target pipefail is greater than our "last_position_to_run",
+          # then it doesn't matter what we return, the recursive parent caller
+          # will not use this value. Just return 0.
+          result_status=0
+        fi
+      else
+        # if we need to return the "last failure",
+        # take the last pipe status, if it failed.
+        if [ "${pipe_status[1]}" -ne 0 ]; then
+          result_status="${pipe_status[1]}"
+        else
+          # if the last pipe status didn't fail,
+          # take whatever the first pipe status is.
+          result_status="${pipe_status[0]}"
+        fi
+      fi
+    fi
+
+    return "$result_status"
+  }
```

Comment:
I find this very involved. The function nesting makes these helpers only visible during execution and they are not prefixed with bats_.

When I thought about this I had in mind something along the lines of (beware untested adhoc code):
```
bats_pipe_recurse () { 
    local return_exit_code=$1
    shift
    # collect left hand side command of (first) pipe
    local first_command=()
    while (( $# > 0 )) && [[ $1 != '|' ]]; do
       first_command+=("$1")
       shift
    done
    if (( $# > 0 )); then # there is a remaining pipe
        shift # consume pipe symbol |
        "${first_command[@]}" | bats_pipe_recurse $(($1 - 1)) "$@" || status=$?
        if (( return_exit_code < 0 )); then
           return $status
        elif (( return_exit_code == 0 )); then
            return "${PIPESTATUS[0]}"
        fi
    else
        # no more pipes -> execute command
        "${first_command[@]}"
    fi
}
```

Reply:
Will take a look into this, just letting you know that I might be a bit slow (maybe a week or 2).

Reply:
@qec-pconner Do you have an updated ETA?

Reply:
Hey, sorry. I've been quite busy. 😓 

I do plan to finish this up, I also have some other fixes (to be other PRs) that I would like to put up. I will likely get to all of this in early December.

Thank you so much, and sorry again for the sudden slow down.

Reply:
> Hey, sorry. I've been quite busy. 😓 

Don't worry, I am busy with other tasks and could take over if necessary. Just tell me before you are working in it, so we don't duplicate our efforts.

> I do plan to finish this up, I also have some other fixes (to be other PRs) that I would like to put up.

Keep 'em coming. :)

Reply:
The end of my current sprint is in sight. Soon™

Reply:
Hope you had a good holiday. Will be picking this back up. Very sorry for all the delays. 😅

Reply:
So I put in the function that you wrote, with some alterations.
I added comments to clarify flows/values.

My approach was trying to have single state and absolute values that did not change on recursion (easier to parse and debug) -- though I think your approach is simpler and more straightforward.

I left in both my functions and your version as well so that they can be seen side-by-side (although execution of my original functions is currently commented out). After settling on one approach, I can remove all the other stuff and clean up in another commit.

Regarding previous comments:
> The function nesting makes these helpers only visible during execution

I purposefully didn't want to expose these "private"/"helper" functions or consider them as part of the API. They (IMO) should be considered implementation details. My functions (`run_command_at_position` and `run_leading_commands_until_position`) depended on variables in parent scope (e.g. `pipe_positions`). I wanted to enforce this dependency (and make it more explicit / clear) by nesting these functions.

> ..and they are not prefixed with bats_.

Again, I wanted to keep them hidden/not exposed. I didn't bother adding this prefix as a byproduct of that, but if you wanted these nested functions to be prefixed as well, I can make that change.

Reply:
I also pushed that other PR that I alluded to (https://github.com/bats-core/bats-core/pull/691).